### ISSUE TITLE
Volatile wrapper has significant undefined behavior

### ISSUE THREAD

**Issue by gnzlbg** [gz/rust-driverkit#1]
Hi, please don't be offended by this. 

I'd like to know: 

* Why did you decided to write your own `Volatile<T>` wrapper ? Did you try to search if this was already done in crates.io ? If so, did you evaluate the alternatives and they did not suited you? Why?

* Why did you decide to use `core::intrinsics` volatile methods instead of the `core::ptr::read_volatile/write_volatile` methods ? 

* Why does your volatile wrapper need to be `repr(packed)` ? It only has one field. 

---

The `write` method writes through a shared reference `&self.foo as *const _ as *mut _`, which is undefined behavior. I suppose this was just an oversight and you aimed to write `&mut self.foo`.

**Comment by gz** [comment]
Hi, not offended, or should I be ;)?

This code was written a long time ago and put into git later so it could be that there wasn't a volatile crate or ptr::{read,write}_volatile methods. But I'm not sure. The repr(packed) and write implementation are obviously bugs.

The crate is not as well maintained as it should be. I think it is quite useful though, I'm currently using this internally to run user-space drivers on three different OSs. Do you feel like submitting a patch that fixes these issues? I'm also not against using an external crate for volatile as long as it doesn't break no_std and is light-weight enough.

**Comment by gnzlbg** [comment]
> This code was written a long time ago and put into git later so it could be that there wasn't a volatile crate or ptr::{read,write}_volatile methods.

That makes sense!

> I'm also not against using an external crate for volatile as long as it doesn't break no_std and is light-weight enough.

There are a couple of crates for this. The only implementation I've checked is the implementation of https://github.com/embed-rs/volatile , but I don't know if you are constrained by any Rust versions. Implementing this does not require much code, but it would be nice to have only one crate to do this, so that if we detect bugs, we only have to find them in one place.

> Do you feel like submitting a patch that fixes these issues?

I'd love to, but I am not sure if I have enough time. I just landed here through a crater run that regressed (aka broke) the crate (*) and wanted to understand why things were done the way they were and give you a heads up about these potential issues.

(*) we are trying to error on some types of undefined behavior at compile time - in this case, the check thought that there was UB because of the reference to the packed struct field, but since the struct only has one field, this field should actually always be properly aligned, so I guess we have to improve the check 😆 .
